lib/xcode/install.rb

```diff
@@ -6,35 +6,106 @@
 require 'rubygems/version'
 require 'xcode/install/command'
 require 'xcode/install/version'
+require 'shellwords'
+require 'open3'
+require 'fileutils'
 
 module XcodeInstall
   CACHE_DIR = Pathname.new("#{ENV['HOME']}/Library/Caches/XcodeInstall")
   class Curl
     COOKIES_PATH = Pathname.new('/tmp/curl-cookies.txt')
 
-    def fetch(url, directory = nil, cookies = nil, output = nil, progress = true)
+    # @param url: The URL to download
+    # @param directory: The directory to download this file into
+    # @param cookies: Any cookies we should use for the download (used for auth with Apple)
+    # @param output: A PathName for where we want to store the file
+    # @param progress: parse and show the progress?
+    # @param progress_block: A block that's called whenever we have an updated progress %
+    #                        the parameter is a single number that's literally percent (e.g. 1, 50, 80 or 100)
+    def fetch(url: nil,
+              directory: nil,
+              cookies: nil,
+              output: nil,
+              progress: nil,
+              progress_block: nil)
       options = cookies.nil? ? [] : ['--cookie', cookies, '--cookie-jar', COOKIES_PATH]
-      # options << ' -vvv'
 
       uri = URI.parse(url)
       output ||= File.basename(uri.path)
       output = (Pathname.new(directory) + Pathname.new(output)) if directory
 
+      # Piping over all of stderr over to a temporary file
+      # the file content looks like this:
+      #  0 4766M    0 6835k    0     0   573k      0  2:21:58  0:00:11  2:21:47  902k
+      # This way we can parse the current %
+      # The header is
+      #  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+      #
+      # Discussion for this on GH: https://github.com/KrauseFx/xcode-install/issues/276
+      # It was not easily possible to reimplement the same system using built-in methods
+      # especially when it comes to resuming downloads
+      # Piping over stderror to Ruby directly didn't work, due to the lack of flushing
```

Comment:
I recently lost a bunch of time with this same problem. I wonder if there is a way to set the buffer size to 0 on the file descriptor.